File: bindings/go/perf/multi/perf_multi_socket_reqrep.go
package main

import (
	"context"
	"errors"
	"fmt"
	"os"
	"strconv"
	"time"

	zlink "zlink.systems/zlink"
	"zlink.systems/zlink/perf/internal/perfcommon"
)

type multiReqRepClient struct {
	target  zlink.SocketTarget
	monitor *zlink.SocketMonitor
	request func() zlink.RequestOp
}

func runMultiDealerRouterReqRepServer(cfg multiConfig) { runMultiSocketReqRepServer(cfg, false) }
func runMultiRouterRouterReqRepServer(cfg multiConfig) { runMultiSocketReqRepServer(cfg, true) }

func runMultiSocketReqRepServer(cfg multiConfig, hasRoutingID bool) {
	serverCtx, err := perfcommon.NewMultiServerContext()
	perfcommon.Must(err)
	defer serverCtx.Close()
	server, err := serverCtx.RouterSocket()
	perfcommon.Must(err)
	defer server.Close()
	if hasRoutingID {
		perfcommon.Must(server.SetRoutingID(zlink.NewRoutingID([]byte("SERVER"))))
		perfcommon.Must(server.SetMandatory(true))
	}
	perfcommon.Must(perfcommon.ConfigureTLSServer(server, cfg.transport))
	perfcommon.ApplyMultiHWM(server, cfg.pattern)
	perfcommon.ApplyMultiBenchmarkSocketOptions(server, cfg.transport)
	endpoint := perfcommon.BindAndResolveEndpoint(server, cfg.transport, "perf-multi-socket-reqrep")
	flushControlLine("READY,%s", endpoint)
	stop := waitForStopAsync()
	poller := perfcommon.NewSocketPoller(server, perfcommon.ZLinkPollIn)
	defer poller.Close()
	events := make([]zlink.PollEvent, 1)
	autoHWMPrinted := false
	for {
		select {
		case <-stop:
			return
		default:
		}
		event, pollErr := perfcommon.WaitPollerOne(poller, events, 100*time.Millisecond)
		if pollErr != nil {
			if perfcommon.IsTransient(pollErr) {
				continue
			}
			perfcommon.Must(pollErr)
		}
		if event == nil || event.Revents&perfcommon.ZLinkPollIn == 0 {
			continue
		}
		var received zlink.Received
		for {
			select {
			case <-stop:
				return
			default:
			}
			ok, recvErr := server.Recv(&received, zlink.RecvFlagsDontWait)
			if recvErr != nil {
				if perfcommon.IsTransient(recvErr) {
					break
				}
				perfcommon.Must(recvErr)
			}
			if !ok {
				break
			}
			if !autoHWMPrinted {
				perfcommon.Must(serverCtx.RecalculateAutoHwm())
				perfcommon.PrintMultiSocketAutoHWMDetail(
					server, nil, cfg.pattern, cfg.transport, "server", "endpoint", zlink.SocketTypeRouter, cfg.msgSize,
				)
				autoHWMPrinted = true
			}
			payload, payloadErr := perfcommon.MeasurementPayload(received.Parts())
			perfcommon.Must(payloadErr)
			if _, valid := received.ReplyToken(); !valid || !received.HasRoutingID() {
				_ = received.Close()
				perfcommon.Must(fmt.Errorf("multi reqrep request missing routing ID or reply token"))
			}
			reply := received.Reply().Message(payload)
			var tail *zlink.Message
			if perfcommon.MeasurementPartCount() == 2 {
				tail = perfcommon.NewMessageWithSize(0)
				reply = reply.Message(tail)
			}
			replyErr := reply.Submit(context.Background())
			if tail != nil {
				_ = tail.Close()
			}
			_ = received.Close()
			if replyErr != nil {
				select {
				case <-stop:
					return
				default:
					perfcommon.Must(replyErr)
				}
			}
		}
	}
}

func runMultiDealerRouterReqRepClient(cfg multiConfig, endpoint string) {
	ctx, err := perfcommon.NewMultiClientContext()
	perfcommon.Must(err)
	defer ctx.Close()
	clients := make([]multiReqRepClient, 0, cfg.clients)
	for i := 0; i < cfg.clients; i++ {
		socket, socketErr := ctx.DealerSocket()
		if socketErr != nil {
			perfcommon.Must(fmt.Errorf("multi reqrep create dealer[%d]: %w", i, socketErr))
		}
		monitor := perfcommon.OpenMonitor(socket)
		if err := perfcommon.ConfigureTLSClient(socket, cfg.transport); err != nil {
			perfcommon.Must(fmt.Errorf("multi reqrep configure dealer[%d]: %w", i, err))
		}
		if err := socket.SetRoutingID(zlink.NewRoutingID([]byte(fmt.Sprintf("dealer-req-%06d", i)))); err != nil {
			perfcommon.Must(fmt.Errorf("multi reqrep set dealer routing id[%d]: %w", i, err))
		}
		if err := socket.Connect(endpoint); err != nil {
			perfcommon.Must(fmt.Errorf("multi reqrep connect dealer[%d]: %w", i, err))
		}
		clients = append(clients, multiReqRepClient{target: socket, monitor: monitor, request: socket.Request})
	}
	runMultiReqRepClients(cfg, ctx, clients)
}

func runMultiRouterRouterReqRepClient(cfg multiConfig, endpoint string) {
	ctx, err := perfcommon.NewMultiClientContext()
	perfcommon.Must(err)
	defer ctx.Close()
	serverID := zlink.NewRoutingID([]byte("SERVER"))
	clients := make([]multiReqRepClient, 0, cfg.clients)
	for i := 0; i < cfg.clients; i++ {
		socket, socketErr := ctx.RouterSocket()
		perfcommon.Must(socketErr)
		monitor := perfcommon.OpenMonitor(socket)
		perfcommon.Must(perfcommon.ConfigureTLSClient(socket, cfg.transport))
		perfcommon.Must(socket.SetRoutingID(zlink.NewRoutingID([]byte(fmt.Sprintf("router-req-%06d", i)))))
		perfcommon.Must(socket.SetMandatory(true))
		perfcommon.Must(socket.SetConnectRoutingID(serverID))
		perfcommon.Must(socket.Connect(endpoint))
		clientSocket := socket
		clients = append(clients, multiReqRepClient{
			target:  socket,
			monitor: monitor,
			request: func() zlink.RequestOp { return clientSocket.Request(serverID) },
		})
	}
	runMultiReqRepClients(cfg, ctx, clients)
}

// Go's public request terminal blocks through reply completion, so each
// logical request runs in its own goroutine while this loop owns completion
// progress for every requester socket.
func runMultiReqRepClients(cfg multiConfig, clientCtx *zlink.Context, clients []multiReqRepClient) {
	defer func() {
		for i := range clients {
			_ = clients[i].monitor.Close()
			switch socket := clients[i].target.(type) {
			case *zlink.DealerSocket:
				_ = socket.Close()
			case *zlink.RouterSocket:
				_ = socket.Close()
			}
		}
	}()
	for i := range clients {
		perfcommon.WaitConnectedWithTimeout(perfcommon.MultiReadyTimeout(), clients[i].monitor)
	}
	// PERF_MULTI_TEST_POLICY § 1.6: recalculate after target connections ready.
	if err := clientCtx.RecalculateAutoHwm(); err != nil {
		perfcommon.Must(fmt.Errorf("multi reqrep client auto-HWM recalc: %w", err))
	}
	timeout := 200 * time.Millisecond
	if value, err := strconv.Atoi(os.Getenv("PERF_MULTI_REQREP_TIMEOUT_MS")); err == nil && value > 0 {
		timeout = time.Duration(value) * time.Millisecond
	}
	result := runMultiReqRepWindow(cfg, clients, activeDeadline(cfg.duration), timeout)
	if len(clients) > 0 {
		socketType := zlink.SocketTypeDealer
		if _, ok := clients[0].target.(*zlink.RouterSocket); ok {
			socketType = zlink.SocketTypeRouter
		}
		perfcommon.PrintMultiSocketAutoHWMDetail(
			clients[0].target, clients[0].monitor, cfg.pattern, cfg.transport, "client", "endpoint", socketType, cfg.msgSize,
		)
	}
	printMultiResult(cfg, result)
	flushControlLine("CLIENT_DONE,%d", cfg.msgSize)
	// The server can still be finishing a reply. Keep its completion targets
	// alive until the runner has stopped the server, then close the sockets.
	waitForStopToken()
}

func runMultiReqRepWindow(cfg multiConfig, clients []multiReqRepClient, window perfcommon.BenchmarkWindow, timeout time.Duration) perfcommon.Result {
	stats := perfcommon.NewMultiStats()
	type counts struct{ attempts, failures, timeouts uint64 }
	countsByClient := make([]counts, len(clients))
	type admission struct {
		client int
		err    error
	}
	type completion struct {
		client      int
		parts       []*zlink.Message
		err         error
		completedAt time.Time
		completedNs int64
	}
	completed := make(chan completion, len(clients))
	admitted := make(chan admission, len(clients))
	available := make([]bool, len(clients))
	for index := range available {
		available[index] = true
	}
	totalOutstanding := 0
	pendingAdmissions := 0

	submitOne := func(index int) {
		countsByClient[index].attempts++
		payload := perfcommon.PreparePayload(cfg.msgSize)
		perfcommon.StampWindowPayload(payload, window.ActiveAt)
		submit := clients[index].request().Bytes(payload)
		if perfcommon.MeasurementPartCount() == 2 {
			submit = submit.Bytes(nil)
		}
		submission, err := submit.Timeout(timeout).Submit(context.Background())
		if err != nil {
			countsByClient[index].failures++
			perfcommon.Must(err)
		}
		totalOutstanding++
		go func() {
			parts, err := submission.Reply(context.Background())
			completed <- completion{
				client: index, parts: parts, err: err,
				completedAt: time.Now(), completedNs: perfcommon.MonotonicNowNs(),
			}
		}()
		if submission.Result() == zlink.SubmitBackpressured {
			available[index] = false
			pendingAdmissions++
			go func() {
				admitted <- admission{client: index, err: submission.Admitted(context.Background())}
			}()
		}
	}

	processCompletion := func(done completion) {
		totalOutstanding--
		count := &countsByClient[done.client]
		if done.err != nil {
			count.failures++
			var requestErr *zlink.RequestError
			zlink.MultipartClose(done.parts)
			if errors.As(done.err, &requestErr) && requestErr.Result == zlink.RequestTimedOut {
				count.timeouts++
				return
			}
			// BACKPRESSURED with a WRITABLE token is retained and resumed by
			// the binding terminal. Any error returned to the runner is fatal.
			perfcommon.Must(done.err)
		}
		reply, payloadErr := perfcommon.MeasurementPayload(done.parts)
		if payloadErr == nil && done.completedAt.Before(window.StopAt) {
			if latency, valid := perfcommon.LatencyNsFromMessageAt(
				reply, cfg.msgSize, perfcommon.PhaseActive, done.completedNs); valid {
				stats.AddLatencyNs(latency / 2.0)
			}
		}
		zlink.MultipartClose(done.parts)
	}

	drainReady := func() bool {
		progressed := false
		for {
			select {
			case ready := <-admitted:
				pendingAdmissions--
				if ready.err != nil {
					perfcommon.Must(ready.err)
				}
				available[ready.client] = true
				progressed = true
			case done := <-completed:
				processCompletion(done)
				progressed = true
			default:
				return progressed
			}
		}
	}

	waitForProgress := func(deadline time.Time) {
		remaining := time.Until(deadline)
		if remaining <= 0 {
			return
		}
		select {
		case ready := <-admitted:
			pendingAdmissions--
			if ready.err != nil {
				perfcommon.Must(ready.err)
			}
			available[ready.client] = true
		case done := <-completed:
			processCompletion(done)
		case <-time.After(remaining):
		}
	}

	for time.Now().Before(window.StopAt) {
		submitted := false
		for index := range clients {
			if !time.Now().Before(window.StopAt) {
				break
			}
			if !available[index] {
				continue
			}
			submitOne(index)
			submitted = true
		}
		progressed := drainReady()
		if !submitted && !progressed {
			waitForProgress(window.StopAt)
		}
	}

	drainTimeout := max(1000*time.Millisecond, timeout*4)
	if value, err := strconv.Atoi(os.Getenv("PERF_MULTI_REQREP_DRAIN_TIMEOUT_MS")); err == nil && value > 0 {
		drainTimeout = time.Duration(value) * time.Millisecond
	}
	drainDeadline := time.Now().Add(drainTimeout)
	for totalOutstanding > 0 || pendingAdmissions > 0 {
		progressed := drainReady()
		if totalOutstanding == 0 && pendingAdmissions == 0 {
			break
		}
		if !time.Now().Before(drainDeadline) {
			perfcommon.Must(fmt.Errorf("multi reqrep completion drain timed out with %d outstanding", totalOutstanding))
		}
		if !progressed {
			waitForProgress(drainDeadline)
		}
	}
	var total counts
	for _, count := range countsByClient {
		total.attempts += count.attempts
		total.failures += count.failures
		total.timeouts += count.timeouts
	}
	fmt.Fprintf(os.Stderr, "REQREP_COUNTS,%s,%s,%d,attempts=%d,failures=%d,timeouts=%d\n",
		cfg.pattern, cfg.transport, cfg.msgSize, total.attempts, total.failures, total.timeouts)
	return stats.Snapshot(cfg.duration, cfg.msgSize)
}

File: bindings/rust/perf/multi/src/perf_multi_socket_reqrep.rs
//! Shared multi-socket request/reply benchmark implementation.

use crate::common;
use std::future::Future;
use std::io::{self, BufRead};
use std::pin::Pin;
use std::sync::{
    Arc,
    atomic::{AtomicBool, Ordering},
};
use std::thread;
use std::time::{Duration, Instant};
use zlink::{
    DealerSocket, Message, POLLCOMPLETION, POLLIN, PollEvent, Poller, RecvFlags, RecvResult,
    RouterSocket, RoutingId, SubmitResult, ZlinkError,
};

const SERVER_ROUTING_ID: &[u8] = b"SERVER";

#[derive(Clone, Copy)]
pub struct ReqRepConfig {
    pattern: &'static str,
    router_clients: bool,
    server_has_routing_id: bool,
}

impl ReqRepConfig {
    pub const fn dealer_router() -> Self {
        Self {
            pattern: "MULTI_DEALER_ROUTER_REQREP",
            router_clients: false,
            server_has_routing_id: false,
        }
    }

    pub const fn router_router() -> Self {
        Self {
            pattern: "MULTI_ROUTER_ROUTER_REQREP",
            router_clients: true,
            server_has_routing_id: true,
        }
    }
}

enum RequestClientSocket {
    Dealer(DealerSocket),
    Router {
        socket: RouterSocket,
        target: RoutingId,
    },
}

type RequestTask = Pin<Box<dyn Future<Output = Result<Vec<Message>, ZlinkError>> + Send>>;

impl RequestClientSocket {
    fn request_task(&self, payload: Message, timeout: Duration) -> RequestTask {
        let operation = match self {
            Self::Dealer(socket) => socket.request(),
            Self::Router { socket, target } => socket.request(target),
        }
        .message(payload);
        let future = if common::measurement_part_count() == 2 {
            operation
                .message(Message::try_from(&[] as &[u8]).expect("empty request tail"))
                .timeout(timeout)
                .submit()
        } else {
            operation.timeout(timeout).submit()
        };

        Box::pin(future)
    }
}

pub fn run_server(config: ReqRepConfig) {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();
    let ctx = common::perf_server_context();
    let router = ctx.router_socket().expect("router");
    if config.server_has_routing_id {
        router
            .set_routing_id(&RoutingId::from(SERVER_ROUTING_ID))
            .expect("set server routing id");
    }
    common::apply_multi_hwm(&router, &settings);
    router
        .common_options()
        .set_send_timeout(Duration::from_millis(settings.send_timeout_ms))
        .expect("send timeout");
    router
        .common_options()
        .set_receive_timeout(Duration::from_millis(settings.receive_timeout_ms))
        .expect("recv timeout");
    if matches!(args.transport.as_str(), "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        common::setup_raw_tls_server(&router, &tls).expect("server tls");
    }

    let Some(bind_endpoint) = common::resolve_server_bind_endpoint(config.pattern, &args.transport)
    else {
        return;
    };
    if let Err(error) = router.bind(&bind_endpoint) {
        if common::handle_transport_setup_error(config.pattern, &args.transport, "bind", error) {
            return;
        }
        panic!("bind: {error}");
    }
    ctx.recalculate_auto_hwm().expect("recalculate auto hwm");
    common::print_ready(&router.last_endpoint().expect("endpoint"));

    let stop = Arc::new(AtomicBool::new(false));
    let stop_reader = Arc::clone(&stop);
    thread::spawn(move || {
        let stdin = io::stdin();
        for line in stdin.lock().lines() {
            match line {
                Ok(line) if matches!(line.trim(), "STOP" | "QUIT") => break,
                Ok(_) => {}
                Err(_) => break,
            }
        }
        stop_reader.store(true, Ordering::Release);
    });

    let poller = Poller::new().expect("poller");
    poller.add_socket(&router, POLLIN, 0).expect("poller add");
    let mut events = vec![PollEvent::default(); 1];
    let mut received = zlink::Received::empty();
    while !stop.load(Ordering::Acquire) {
        let event_count = poller.wait(&mut events, 100).expect("poller wait");
        if stop.load(Ordering::Acquire) {
            break;
        }
        if events[..event_count]
            .iter()
            .all(|event| event.slot != 0 || event.revents & POLLIN == 0)
        {
            continue;
        }

        while !stop.load(Ordering::Acquire) {
            match router.recv(&mut received, RecvFlags::DONT_WAIT) {
                Ok(true) => reply_request(&received),
                Ok(false) => break,
                Err(error) if error.code() == RecvResult::NoData => break,
                Err(error) => panic!("request receive failed: {error}"),
            }
        }
    }
}

fn reply_request(received: &zlink::Received) {
    if received.routing_id().is_none() || received.reply_token().is_none() {
        panic!("request is missing its reply route");
    }
    let payload = common::message_payload(received.parts());
    if payload.is_empty() {
        panic!("request has an invalid measurement envelope");
    }

    // Message::try_clone uses zlink_msg_copy, so the reply owns an independent
    // native reference without copying the payload bytes out to a Vec and back.
    let reply_payload = received.parts()[0]
        .try_clone()
        .expect("clone received reply payload");
    let reply = received.reply().message(reply_payload);
    let result = if common::measurement_part_count() == 2 {
        reply
            .message(Message::try_from(&[] as &[u8]).expect("empty reply tail"))
            .submit()
    } else {
        reply.submit()
    };
    match result {
        Ok(()) => {}
        Err(error)
            if matches!(
                error.code(),
                SubmitResult::NotConnected | SubmitResult::NotFound
            ) => {}
        Err(error) => panic!("request reply failed: {error}"),
    }
}

pub fn run_client(config: ReqRepConfig) {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();
    let ctx = common::perf_client_context();
    let server_routing_id = RoutingId::from(SERVER_ROUTING_ID);
    let mut sockets = Vec::with_capacity(settings.clients);
    let mut monitors = Vec::with_capacity(settings.clients);

    for index in 0..settings.clients {
        let routing_id = RoutingId::from(format!("CLIENT-{index}").as_bytes());
        if config.router_clients {
            let socket = ctx.router_socket().expect("router client");
            common::apply_multi_hwm(&socket, &settings);
            configure_client_options(&socket, &settings);
            socket
                .set_routing_id(&routing_id)
                .expect("set client routing id");
            socket
                .router_options()
                .set_connect_routing_id(&server_routing_id)
                .expect("set connect routing id");
            if matches!(args.transport.as_str(), "tls" | "wss") {
                let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
                common::setup_raw_tls_client(&socket, &tls).expect("client tls");
            }
            let monitor = common::open_connection_ready_monitor(&socket);
            socket.connect(&args.endpoint).expect("connect");
            sockets.push(RequestClientSocket::Router {
                socket,
                target: server_routing_id,
            });
            monitors.push(monitor);
        } else {
            let socket = ctx.dealer_socket().expect("dealer client");
            common::apply_multi_hwm(&socket, &settings);
            configure_client_options(&socket, &settings);
            socket
                .set_routing_id(&routing_id)
                .expect("set client routing id");
            if matches!(args.transport.as_str(), "tls" | "wss") {
                let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
                common::setup_raw_tls_client(&socket, &tls).expect("client tls");
            }
            let monitor = common::open_connection_ready_monitor(&socket);
            socket.connect(&args.endpoint).expect("connect");
            sockets.push(RequestClientSocket::Dealer(socket));
            monitors.push(monitor);
        }
    }

    let ready_timeout = common::resolve_multi_connect_ready_timeout();
    for monitor in &mut monitors {
        common::wait_monitor_ready(monitor, ready_timeout, "multi request client");
    }
    drop(monitors);
    ctx.recalculate_auto_hwm().expect("recalculate auto hwm");

    let completion_poller = Poller::new().expect("request completion poller");
    for (slot, socket) in sockets.iter().enumerate() {
        match socket {
            RequestClientSocket::Dealer(socket) => completion_poller
                .add_socket(socket, POLLCOMPLETION, slot)
                .expect("request completion poller add"),
            RequestClientSocket::Router { socket, .. } => completion_poller
                .add_socket(socket, POLLCOMPLETION, slot)
                .expect("request completion poller add"),
        }
    }
    let mut completion_events = vec![PollEvent::default(); sockets.len().max(1)];

    let request_timeout = common::resolve_multi_reqrep_timeout();
    let payload_size = args.msg_size.max(common::HEADER_SIZE);
    let active_deadline = Instant::now() + Duration::from_secs(settings.duration_seconds);
    let mut sequences = vec![1u64; sockets.len()];
    let mut requests = common::ConcurrentTasks::<RequestTask>::new(0);
    let mut latency = common::LatencyStats::new();

    // There is deliberately no per-socket or global application window. Each
    // request Future owns any refused payload and waits for its own WRITABLE
    // token before resubmitting; admitted requests remain in flight until the
    // Core REQUEST completion arrives.
    while Instant::now() < active_deadline {
        for (socket_index, socket) in sockets.iter().enumerate() {
            if Instant::now() >= active_deadline {
                break;
            }
            let sequence = sequences[socket_index];
            sequences[socket_index] = sequence.wrapping_add(1);
            let mut payload = Message::with_size(payload_size).expect("request payload");
            common::encode_header(
                payload.data_mut(),
                common::PHASE_ACTIVE,
                args.msg_size as u32,
                sequence,
            );
            requests.push(socket.request_task(payload, request_timeout));
        }

        let ready = requests.poll_ready();
        let progressed = !ready.is_empty();
        for (_, completion) in ready {
            process_completion(completion, args.msg_size, active_deadline, &mut latency);
        }
        if Instant::now() < active_deadline {
            // WRITABLE and REQUEST records share the socket completion lane.
            // Even after executor progress, a zero-time wait drains that lane;
            // an idle turn blocks until Core wakes it or the phase ends.
            let wait_ms = if progressed {
                0
            } else {
                common::poll_timeout_until(active_deadline)
            };
            completion_poller
                .wait(&mut completion_events, wait_ms)
                .expect("request completion wait");
        }
    }

    let drain_deadline =
        Instant::now() + common::resolve_multi_reqrep_drain_timeout(request_timeout);
    while requests.any_pending() && Instant::now() < drain_deadline {
        let ready = requests.poll_ready();
        let progressed = !ready.is_empty();
        for (_, completion) in ready {
            process_completion(completion, args.msg_size, active_deadline, &mut latency);
        }
        if requests.any_pending() {
            let wait_ms = if progressed {
                0
            } else {
                common::poll_timeout_until(drain_deadline)
            };
            completion_poller
                .wait(&mut completion_events, wait_ms)
                .expect("request completion drain wait");
        }
    }
    for (_, completion) in requests.poll_ready() {
        process_completion(completion, args.msg_size, active_deadline, &mut latency);
    }
    assert!(
        !requests.any_pending(),
        "request completion drain timed out"
    );

    common::print_result(
        config.pattern,
        &args.transport,
        args.msg_size,
        settings.duration_seconds,
        &latency.finish(),
    );
}

fn configure_client_options<S>(socket: &S, settings: &common::MultiSettings)
where
    S: ClientCommonOptions,
{
    socket
        .set_send_timeout(Duration::from_millis(settings.send_timeout_ms))
        .expect("send timeout");
    socket
        .set_receive_timeout(Duration::from_millis(settings.receive_timeout_ms))
        .expect("recv timeout");
}

trait ClientCommonOptions {
    fn set_send_timeout(&self, timeout: Duration) -> Result<(), ZlinkError>;
    fn set_receive_timeout(&self, timeout: Duration) -> Result<(), ZlinkError>;
}

macro_rules! impl_client_common_options {
    ($($socket:ty),+ $(,)?) => {
        $(
            impl ClientCommonOptions for $socket {
                fn set_send_timeout(&self, timeout: Duration) -> Result<(), ZlinkError> {
                    Ok(self.common_options().set_send_timeout(timeout)?)
                }

                fn set_receive_timeout(&self, timeout: Duration) -> Result<(), ZlinkError> {
                    Ok(self.common_options().set_receive_timeout(timeout)?)
                }
            }
        )+
    };
}

impl_client_common_options!(DealerSocket, RouterSocket);

fn process_completion(
    outcome: Result<Vec<Message>, ZlinkError>,
    message_size: usize,
    active_deadline: Instant,
    latency: &mut common::LatencyStats,
) {
    match outcome {
        Ok(parts) => {
            if Instant::now() < active_deadline {
                let payload = common::message_payload(&parts);
                common::record_active_rtt_latency(payload, message_size, latency);
            }
        }
        // Request-domain completion failures are terminal outcomes; keep
        // driving the active window after accounting for them.
        Err(ZlinkError::Request(_)) => {}
        Err(error) => panic!("request failed: {error}"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn official_configs_keep_distinct_request_roles() {
        let dealer_router = ReqRepConfig::dealer_router();
        assert_eq!(dealer_router.pattern, "MULTI_DEALER_ROUTER_REQREP");
        assert!(!dealer_router.router_clients);
        assert!(!dealer_router.server_has_routing_id);

        let router_router = ReqRepConfig::router_router();
        assert_eq!(router_router.pattern, "MULTI_ROUTER_ROUTER_REQREP");
        assert!(router_router.router_clients);
        assert!(router_router.server_has_routing_id);
    }
}
